From 50ceefa5a18049e4004fec38de0038434e452134 Mon Sep 17 00:00:00 2001
From: Limeng <Meng.Li@windriver.com>
Date: Sun, 21 Jun 2020 13:05:10 +0800
Subject: [PATCH 37/37] driver: gpu: dup: stop frame generator if it is in
 enabled status

When run kexec/dump feature, it is need to stop frame generator.
Because it is in enabled status when the second kernel boots up.
The delay time is used to make sure stop frame generator completely.
The delay time is related with the size of image resolution. I don't
get the precise reference value from datasheet. So, the value of
delay time is the 2 times of experiment value.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/dpu/dpu-crtc.c | 5 +++++
 drivers/gpu/imx/dpu/dpu-framegen.c | 5 +++++
 include/video/dpu.h                | 1 +
 3 files changed, 11 insertions(+)

diff --git a/drivers/gpu/drm/imx/dpu/dpu-crtc.c b/drivers/gpu/drm/imx/dpu/dpu-crtc.c
index ee8476cfcfae..45d3292ad37f 100644
--- a/drivers/gpu/drm/imx/dpu/dpu-crtc.c
+++ b/drivers/gpu/drm/imx/dpu/dpu-crtc.c
@@ -1135,6 +1135,11 @@ static int dpu_crtc_init(struct dpu_crtc *dpu_crtc,
 		return ret;
 	}
 
+	if (framegen_is_enabled(dpu_crtc->fg)) {
+		framegen_disable(dpu_crtc->fg);
+		mdelay(500);
+	}
+
 	plane_grp->res.fg[stream_id] = dpu_crtc->fg;
 	dpu_crtc->plane[0] = dpu_plane_init(drm, 0, stream_id, plane_grp,
 					DRM_PLANE_TYPE_PRIMARY);
diff --git a/drivers/gpu/imx/dpu/dpu-framegen.c b/drivers/gpu/imx/dpu/dpu-framegen.c
index ff71bbcef502..472075cb8711 100644
--- a/drivers/gpu/imx/dpu/dpu-framegen.c
+++ b/drivers/gpu/imx/dpu/dpu-framegen.c
@@ -262,6 +262,11 @@ void framegen_enable(struct dpu_framegen *fg)
 }
 EXPORT_SYMBOL_GPL(framegen_enable);
 
+bool framegen_is_enabled(struct dpu_framegen *fg)
+{
+	return dpu_fg_read(fg, FGENABLE);
+}
+
 void framegen_disable(struct dpu_framegen *fg)
 {
 	dpu_fg_write(fg, 0, FGENABLE);
diff --git a/include/video/dpu.h b/include/video/dpu.h
index 1657986c70a2..94532f60587a 100644
--- a/include/video/dpu.h
+++ b/include/video/dpu.h
@@ -625,6 +625,7 @@ void dpu_fw_put(struct dpu_fetchunit *fu);
 
 /* Frame Generator Unit */
 struct dpu_framegen;
+bool framegen_is_enabled(struct dpu_framegen *fg);
 void framegen_enable(struct dpu_framegen *fg);
 void framegen_disable(struct dpu_framegen *fg);
 void framegen_enable_pixel_link(struct dpu_framegen *fg);
-- 
2.17.1

